[jello] Week2 pair review

hjsong333 opened the pull request:
## 주요 작업
- 코드 줄이기
- Express 서버
- 비동기로 가져온 데이터로 렌더링

## 고민
- 특정 조건에서만 클래스 이름을 추가하는 방법(코드 줄이기의 일환)
- 비동기 처리 후 렌더링

hjsong333 committed:
refactor: remove 2nd parameter of invoke function

- The way to call updateProps() of passed component is bad for re-render of multiple components.
- With the previous way, component knows all the components to render.

hjsong333 committed:
chore: install Express to serve data

hjsong333 committed:
chore: install cors package to handle CORS issue

- Because of different origin(url)

hjsong333 committed:
feat: render GridView with fetched data

- Render GridView with empty data at first.
- After fetch image data, change state of images.
- Rerender GridView with changed state.

hjsong333 committed:
fix: displayed subscription in absence of image

- If `data-id` of image is empty, don't render subscription cover.

hjsong333 committed:
chore: prepare for serving headline news data

- Get headline news data by crawling.
- Make route for headline news.

hjsong333 committed:
feat: display news in NewsBar with fetched data

- First, render empty layout of NewsBar.
- After fetch news data, update text and url of NewsBar.

crongro commented:
젤로도 생성자 코드가 너무~ 길어요. 속성을 초기화 하는것 이외에는 모두 메서드로 분리해서 구현해보세요.

qkdflrgs commented on src/components/main/GridView.ts:
```diff
@@ -88,41 +99,51 @@ export default class GridView {
+    if (this.props.isHover !== isHover || this.props.hoverIndex !== hoverIndex) {
+      this.renderSubscriptionCover(isHover, hoverIndex, props.subscriptionInfo);
+    }
 
-    if (this.page === page) {
-      this.renderSubscriptionCover(isHover, hoverIndex, subscriptionInfo);
-      return;
+    if (this.props.page !== page || this.props.imgs.length !== imgs.length) {
+      this.renderCurrentPage(imgs, page);
```
현재는 모든 속성을 하나씩 비교하면서 상태 변경을 하고 있는데, 상태 변경을 감지하고 필요한 부분만 업데이트하면 성능적인 측면이나 유지보수 측면에서 좋지 않을까 싶습니다!

qkdflrgs commented on src/types.ts:
```diff
@@ -32,6 +44,15 @@ type GridInfo = {
+  newsUrl: string;
+};
+
+type HeadlineInfo = {
+  news: HeadlineNews[];
+  leftIndex: number;
+  rightIndex: number;
+};
```
타입 별칭(Type Aliases)처럼 객체 형태에만 사용되는 인터페이스(interface)라는 친구가 있는데 class처럼 상속도 가능하고 확장성이 좋은 장점이 있습니다! 기회가 된다면 한번 맛보시는 것도 좋을 것 같습니다 :)

[인터페이스:타입스크립트 핸드북](https://joshua1988.github.io/ts/guide/interfaces.html)

hjsong333 committed:
chore: clean up server code and files

- Remove server from express-generator.
- Rebuild server with express, cors.
- Prepare scripts to run client and server simultaneously.
- Change process for each API as simple as possible.

hjsong333 committed:
refactor: clean up constructors of components

- Init properties first.
- Init necessary elements and append elements.
- Render or update view by data from props.
- Set event at the end of constructor.

hjsong333 committed:
chore: remove db directory and json file

hjsong333 committed:
feat: render headline news with fetched data.

- Prepare current and next news element for rolling animation.
- Rename `RollingNews` to `NewsRoller`.

hjsong333 committed:
feat: headline roll up animation

- Implemented by transition.
- Both of news bars roll at the same time.

hjsong333 committed:
feat: roll up headline news independently

- States(index, HeadlineNews) for NewsRoller are contained in component.
- Rolling starts by NewsBar component(parent of NewsRoller).
- Activate/deactivate rolling animation is executed by transition(`style.roll_up`).
